sdk/Lusid.Sdk.Tests/Utilities/PollyApiRetryHandlerTests.cs

```diff
@@ -0,0 +1,409 @@
+﻿using System;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using Lusid.Sdk.Api;
+using Lusid.Sdk.Client;
+using Lusid.Sdk.Utilities;
+using NUnit.Framework;
+using Polly;
+using RestSharp;
+
+namespace Lusid.Sdk.Tests.Utilities
+{
+    [TestFixture]
+    // For further Polly functionality reference, check out this page: https://github.com/App-vNext/Polly
+    public class PollyApiRetryHandlerTests
+    {
+        private LusidApiFactory _apiFactory;
+        private readonly Policy<IRestResponse> _initialRetryPolicy = RetryConfiguration.RetryPolicy;
+        private HttpListener _httpListener;
+        private const string ListenerUriPrefix = "http://localhost:4444/";
+
+        [SetUp]
+        public void SetUp()
+        {
+            _httpListener = new HttpListener();
+            _httpListener.Prefixes.Add(ListenerUriPrefix);
+
+            var testApiConfig = TestLusidApiFactoryBuilder.CreateApiConfiguration(null);
+            testApiConfig.ApiUrl = ListenerUriPrefix;
+
+            _apiFactory = new LusidApiFactory(testApiConfig);
+
+            _httpListener.Start();
+        }
+
+        [Test]
+        public void CallGetPortfoliosApi_WhenHttpStatusIs400AndRetryConditionIsNotSatisfied_ThrowsApiExceptionWithoutRetry()
+        {
+            // It should not retry on codes >400 by default, unless specified
+            const int expectedStatusCode = 400;
+            // Add the next response returned by api
+            AddMockHttpResponseToQueue(_httpListener, statusCode: expectedStatusCode, responseContent: "{\"some\": \"JsonResponseHere\"}");
+            var retryCount = 0;
+            RetryConfiguration.RetryPolicy =
+                Policy.Wrap(
+                    PollyApiRetryHandler.DefaultFallbackPolicy,
+                    Policy
+                        .HandleResult<IRestResponse>(PollyApiRetryHandler.GetInternalExceptionRetryCondition)
+                        .Retry(retryCount: 3, onRetry: (result, i) => retryCount++)
+                );
```

Comment:
As you are using both the `DefaultFallbackPolicy` and `GetInternalExceptionRetryCondition` from the `PollyApiRetryHandler` why not just use `RetryConfiguration.RetryPolicy = PollyApiRetryHandler.InternalExceptionRetryPolicyWithFallback;`?

This way you are testing that the `PollyApiRetryHandler` does not retry on a 400 response.

Or are you trying to isolate the `PollyApiRetryHandler.GetInternalExceptionRetryCondition`? If so you can just call this method with the inputs you want and don't even need to make an API call at all.

Reply:
Note this pattern seems to be across multiple tests.

Reply:
I am only isolating PollyApiRetryHandler.GetInternalExceptionRetryCondition as this is the only testable part.

Using the policy directly from PollyApiRetryHandler I cannot test that the number of retries was 0 as far as I'm aware. Maybe there's a way? wondering if there's a way to create a mock from a working object

Reply:
If you want to test the `PollyApiRetryHandler` in its entirety you could pass a counter into `GetHttpResponseHandler` and have it incremented on each invocation.

Reply:
Done x2

Reply:
Sweet